### ISSUE TITLE
Console output should better reflect the build process

### ISSUE THREAD

**Issue by nnethercote** [rust-lang/cargo#8889]
**Describe the problem you are trying to solve**

The console output can be misleading about what's happening in the build. Having lots of sequential `Compiling <crate>` lines makes it seem like compilation is serial, and if a slow-building crate X starts building just before a fast-building crate Y, you'll probably think that Y is the slow-building crate. (I was misled in this way about build times of different crates within rustc for a long time.)

The progress bar helps with this, because it shows which crates are currently being built... but only if you have a sufficiently wide terminal. If you have 80 chars you don't see them at all. If you have 100 chars you might see one or two or three, but any crate with a long name greatly reduces the effectiveness.

Also, linking isn't shown. This means people often blame the compiler for slowness when it's the linker's fault.

**Describe the solution you'd like**

1. Instead of printing `Compiling <crate>` when a crate starts building, print `Compiled <crate> in X.YYs`. That would avoid the problem where slowness to build is blamed on the wrong crate. This is a change of tense in the message, but Cargo already mixes present and past tenses, e.g. with `Downloaded <crate>` and `Downloaded2 3 crates (78.7 KB) in 1.03s` messages.

1. Shrink the width of the progress bar, so the crates being built can be seen even on narrow terminals.

1. Show linking as a distinct step, including the time it took, e.g. `Linked in X.YYs".

**Notes**

My motivation here was comparing console output (which is moderately useful) to `-Ztimings` output (which is amazing), and trying to make the former a bit more like the latter.

**Comment by nnethercote** [comment]
>     2\. Shrink the width of the progress bar, so the crates being built can be seen even on narrow terminals.

#8892 does this.

**Comment by ehuss** [comment]
I don't think it is currently possible to show link time as a separate stat, since `rustc` does compilation+linking together.  It might be possible to have `rustc` spit out a JSON message when it starts linking, and Cargo could catch that to start a new timer.  There might be a tiny delay, but I suspect not enough to affect the timing in a meaningful way.  I think that would be useful (at least for `-Ztimings`).  Would need to ask the compiler team what they think about adding instrumentation like that.

**Comment by nnethercote** [comment]
Maybe Cargo could output "Compiling and linking" (or "Compiled and linked") instead? Less precise, but still clearer than the current output.

**Comment by alexcrichton** [comment]
The intention of the current progress bar was to give an indicator as to what's currently compiling so we don't have to just read the last line and blame that (ideally), but it's true that we never got back to updating the console output otherwise. I'd be a bit wary of indicating the compile time of each crate individually because that could add up to a lot of noise in the current output scheme.

All that being said though, I think it's high time we completely revamp Cargo's output. Even one-short-line-per-package feels unnecessarily verbose today. I'd sort of prefer if we could have the progress bar we currently have but have maybe at most 4-5 other lines indicating "these were the slowest crates to compile" or something like that. Then when Cargo finishes it could erase the output or leave it all there. (or something like that). 

Basically I think the current output has gotten us this far but it could benefit from a complete overhaul in addition to smaller tweaks.

**Comment by joshtriplett** [comment]
@ehuss We've talked about, for other reasons, separating compilation and linking: https://github.com/rust-lang/rust/issues/64191

If we do that in rustc, cargo should use that, and could then reflect that in its output.

**Comment by dtolnay** [comment]
> I think it's high time we completely revamp Cargo's output. Even one-short-line-per-package feels unnecessarily verbose today.

Sharing two recordings for possible inspiration. Both of these build systems' outputs are designed around accurately surfacing where the most time is getting spent in builds with (extremely) high parallelism. The currently running longest running jobs are always at the top.

<br>

![Peek 2021-02-17 21-01](https://user-images.githubusercontent.com/1940490/108307902-9dea2180-7163-11eb-9a4d-269d68d40d9f.gif)

![Peek 2021-02-17 21-02](https://user-images.githubusercontent.com/1940490/108307921-a7738980-7163-11eb-80c0-4844d55a4390.gif)

**Comment by ehuss** [comment]
Thanks @dtolnay for sharing those.  I'm not too familiar with what kind of terminal detection and support is necessary to be able to create multi-line progress indicators like that, especially across a myriad of platforms.  Cargo's current progress bar is extremely primitive, which makes it a little easier to maintain, but is much more limiting.  

I think it would be great to improve the output, I just hope it doesn't require pulling in large dependencies, or require heroic efforts to get it to work across terminals.  Cygwin/msys on Windows has been particularly hard to support, since Cargo doesn't live in the Cygwin universe, it is limited on what it can do.

I'm also curious how warnings and errors get displayed. One problem I struggled with in the current progress bar implementation is that it has to clear the progress bar in order to display a message (and scroll the output up), and then redisplay the progress bar. This causes severe flickering on some terminals (particularly on Windows).  I'm curious how buck and bazel and other tools handle that.

**Comment by camsteffen** [comment]
Minor suggestion. The list of crates in the **`Building`** line should match the order that they are printed in the **`Compiling`** lines. That is, the one that has been running the longest should be first, IIUC.

**Comment by patricksjackson** [comment]
I am interested in working on this, and have already created a hacky prototype build using [crossterm](https://github.com/crossterm-rs/crossterm) that is heavily influenced by buck/bazel. See the recording below.

@alexcrichton @ehuss @dtolnay Since this is pretty old, could you confirm if you'd actually want to take this change? It's a decent chunk of work to do it properly so I don't want to spend too much time on it if you're no longer interested.

[![asciicast](https://asciinema.org/a/nMUGH5T2PiizxwK340n0DTW4M.svg)](https://asciinema.org/a/nMUGH5T2PiizxwK340n0DTW4M)

**Comment by dtolnay** [comment]
If you are interested in something less hacky, check out https://github.com/facebookincubator/superconsole.

Here is some further inspiration from buck2, which uses superconsole for the UI. This is building the same crate as in your asciinema:

![Peek 2022-07-20 01-24](https://user-images.githubusercontent.com/1940490/179934972-f970e8dc-2c68-4203-9434-c511f8e7a4eb.gif)

**Comment by bjorn3** [comment]
I did like to keep the history of crates that have been built already.

**Comment by dtolnay** [comment]
@bjorn3: in buck there is a `buck log what-ran` command for that purpose, which you'd run after the fact to find out information about the last invocation. I like something like that over filling the terminal with dozens to thousands of lines every time. It can also be convenient for scripting; flags for output format (json) and what information to include (timings, crate authors, licenses, number of warnings, …).

**Comment by patricksjackson** [comment]
I understand that bikeshedding for this type of UX change is inevitable, but I don't want to get caught up in that until we get a confirmation of whether or not this change is still desired.

Let me at least clarify the design ideas that I do have so it's a bit easier to envision. I'm of course open for the project leadership to have the final say on the design but my viewpoint is the following:

1. I think it's important to retain a similar design language to the existing Cargo output, with minimal additions to add the extra bits of info that this new design would allow. We shouldn't be trying to match the "look" of these other build tools, instead we should borrow ideas for what information is useful and incorporate that into Cargo's UX.
2. The extra info that we should be adding over the existing progress bar are a count/list of in progress tasks and timings for at least longer tasks. I think we should continue reporting at the same level of verbosity as current Cargo (one 'task' per crate) instead of breaking them out into individual actions like the buck example above, mainly to maintain compatibility with Cargo's current design language.
3. We should drop the scroll of messages for normal interactive usage in favor of using the progress view only, as in my example above. If we do decide to keep them, they should be reporting of _completed_ work instead of reporting the start of work like today. Reporting work _start_ is the job of the progress bar.
4. For warnings or errors from rustc, we should be able to capture the output and print a line like "Compiled XX with warnings:", followed by the warning message, once the crate has finished compiling. Those messages will be displayed above the progress bar and persist upon exit. That will fix the current issues we have with trying to associate the warning and error messages with the right crate, see #8977.
5. The fallback output for "dumb" terminals or CI should be switched so that we report task completion + captured output at the same time. We already detect that and drop the [progress bar](https://github.com/rust-lang/cargo/blob/caf3c37d90081a211a2457f8b5543ae4cddb0142/src/cargo/util/progress.rs#L43-L45) in those cases.

Getting this to work will require a good amount of refactoring to centralize the output reporting. In brief, we have to pipe all the output to a single stateful progress indicator so that we can pretend that Cargo actually uses a single job queue. We'll need to do this so that we can have all of the output UX handled in the same centralized way and avoid the fragmentation that currently exists. I think doing this lets us present the user with the right mental model even if the underlying details don't quite match.

I would plan to do that work first, without any significant changes to the output UX, and then iterate on the design.

If there's a process that you would prefer is followed for a change of this magnitude, please let me know.

**Comment by nnethercote** [comment]
@patricksjackson I'm no Cargo expert, but I think everything you've written is reasonable and well thought out.

**Comment by dtolnay** [comment]
@patricksjackson I think the best way to make fast progress on the user interface design without overburdening the Cargo maintainers with reviews or Cargo users with an unstable experience is going to be by delegating the whole UI to a separate process.

#### Proposal:

- If Cargo finds something called `cargo-tui` on the path (using whatever lookup hierarchy would occur for a command-line invocation of `cargo tui`) then it will spawn an invocation of it for the duration of the cargo process.
- Cargo will feed events via the subprocess's stdin, as one-line JSON objects.
- When Cargo is done, it will close the subprocess's stdin and wait for it to exit.

That's it. We don't need to debate what information should appear in the UI or what the color and boldness and positioning of everything should be. We'll give the ecosystem 1&ndash;2 years to build a suite of great UI options with different tradeoffs, and then discuss how to import or fork one of them to become the out-of-the-box default UI for cargo.

We can kickstart folks by providing a skeleton crate called `cargo-legacy-tui` (with `[[bin]] name = "cargo-tui"`) which exactly reimplements Cargo's current UI. Anyone interested can fork this and build a fancier experience atop it. Users can swap in any new UI by doing `cargo install cargo-fancy-ui-of-your-choice` (again assuming it provides `[[bin]] name = "cargo-tui"`).

For debugging, or for cargo's test suite, something like `ln -s /bin/cat ~/.cargo/bin/cargo-tui` or `ln -s /usr/bin/jq ~/.cargo/bin/cargo-tui` is a cool trick.

**Comment by patricksjackson** [comment]
@dtolnay I'm confused by your suggestion. This feels like a more complicated solution to build that creates more work for the Cargo team and the community. Your proposed solution provides unclear value considering the community isn't asking for such a solution here, so a flexible solution is not useful. Can you clarify what value you see in this solution compared to my proposal? And what other similar tools find it useful to do something like this?

This issue has sat for multiple years with no interest, so I am proposing a straightforward solution both to implement and to utilize, that I'm willing to build.

**Comment by dtolnay** [comment]
Sure thing; I don't have confidence that iterating on the built-in cargo UI in the way you suggest would lead to the best result in the long term, as opposed to just a local optimum that is better on a few axes than the current UI, but otherwise even harder to iterate further and permanently suboptimal in comparison to state-of-the-art build tools.

Closing the gap between Cargo's current build UI and the state-of-the-art in build UI from other tools requires in my opinion bigger experimentation and bigger disruptive advancements than what it would make sense to iterate on in Cargo's own codebase. Performing those design iterations externally, in a way that Cargo users can easily install and provide feedback on without needing to fork all of cargo, will arrive at a better outcome. The reasons it will arrive at a better outcome are:

- faster iteration: after the initial scaffolding in cargo, the rest of the iteration on what information appears and how, and the TUI implementation details, can all happen in a way that is not beholden to Cargo's review process, which is overburdened;

- parallel iteration: different people who are passionate about this space can pursue different designs, with no need for up-front consensus about what design to pursue;

- better revealed preferences from users: it will become clear through number of downloads, number of issues, number of contributors etc what UI is seen as most preferred by users.

To the extent that you've suggested a straightforward solution, it's a solution to doing better than Cargo's current UI but I don't see it as a solution to matching or advancing the state of the art.

**Comment by patricksjackson** [comment]
Right. Thanks for clarifying.

I admit I'm very curious what useful, actionable information a "state-of-the-art build UI" would present that isn't already presented by the likes of buck/bazel, but I'm happy to let someone else figure that out. The only things I can think of are largely aesthetic choices, like better representations of the DAG.

I continue to feel that the best way forward, balancing the apparent level of interest from the community and ROI for the Cargo team, lies in my suggested path. I think it's perfectly fine for Cargo's TUI to move towards a local maxima representing the latest in "SOTA build UIs"

I may continue to work on a more complete patchset with my proposed changes to shop around, but it'll stay on my backburner unless I run out of other projects or the Cargo team indicates interest.

**Comment by dtolnay** [comment]
- Association of diagnostics to which part of the build graph they were emitted from. Currently not even attempted by Cargo.

- *Post hoc* slicing of build failures. Currently Cargo will interleave diagnostics from crates which built concurrently (thankfully at the granularity of whole errors, not individual lines or characters). This can make it hard to focus in on only errors from a particular crate or file, as you are working on fixing that one crate or file during a refactor, because the output relevant to you can be dwarfed by other errors in the rest of the project.

- Error/warning count and/or categorization.

- Real-time deterministic ordering. Cargo's current UI emits diagnostics in whatever random order the build graph got to them. This makes it annoying because there can be an error/warning at the bottom which you attempt to fix, then rebuild and that error/warning appears to have gone away, but in fact it's just higher up in the output this time and some other error/warning is now the bottom one. There is opportunity to do vastly better with this.

- Most-relevant-up-front heuristics. Normally the best strategy for rust errors is to "start at the top". On the current cargo console this can mean scrolling up past thousands or tens of thousands of irrelevant lines each team you rebuild. It's unfortunate that the immediately visible part of the output after the build is most often the most useless.

- "Open in IDE" functionality where you can click an error in console to jump to the relevant source location in your preferred editor.

**Comment by epage** [comment]
Summary of discussion so far (with some of my own ideas thrown in)

Some interests of what to include
- In static "Compiling X" message, more accurate say "Compiled" and include the duration
- Show a linking step
  - Would require support from the compiler:
    - https://github.com/rust-lang/cargo/issues/8889#issuecomment-732536045
    - https://github.com/rust-lang/cargo/issues/8889#issuecomment-733111297
- Show more crates being compiled
  - ~~Idea: shrink progress bar~~ see #8892 
  - Idea: Show each crate on its own line with elapsed time, oldest first (interactive only, otherwise just show the individual statuses)
    - ![Peek 2021-02-17 21-01](https://user-images.githubusercontent.com/1940490/108307902-9dea2180-7163-11eb-9a4d-269d68d40d9f.gif)
    - ![Peek 2021-02-17 21-02](https://user-images.githubusercontent.com/1940490/108307921-a7738980-7163-11eb-80c0-4844d55a4390.gif)
    - ![Peek 2022-07-20 01-32](https://user-images.githubusercontent.com/1940490/179936177-7fd0a80e-7e77-4368-a2e5-4f10cd7b4788.gif)
- one-short-line-per-package feels unnecessarily verbose today
  - Even worse with combined with test output (#5089)
- Some users would like to to still see history
  - Buck splits this out into a `buck log what-ran`
  - If we logged like this, we could also log the rebuild reasons, making it easier to diagnose why a build happened
- Pager support (#11213)
  - See https://crates.io/crates/bacon
- Order output so users have initial error in front of them, then subsequent errors, then warnings (with a pager, this is always at the top, without a pager, it depends)
  - See https://crates.io/crates/bacon
- Absolute minimal non-transient output for "cargo script", see #12207

Implementation
- Cargo is using a bespoke progress bar implementation
  - gix (a dependency) is using [prodash](https://crates.io/crates/prodash)
  - [superconsole](https://github.com/facebookincubator/superconsole) is used by buck2
- Allow faster iteration (and user customization) by making the UX a plugin system
  - https://github.com/rust-lang/cargo/issues/8889#issuecomment-1192159222
  - Requires stable json output from cargo

**Comment by epage** [comment]
We discussed this briefly in the cargo team meeting.  A concern raised about the fancy superconsole output is how universally the ansi escape codes needed are supported and how well we can detect and fallback in that situation.

Currently, `cargo` skips progress bars for quiet, CI, or if `TERM=dumb` ([code](https://github.com/rust-lang/cargo/blob/4f6553ab553ca127ab653bec9f2f65e64735912f/src/cargo/util/progress.rs#L40)).  The CI check was added in #6281 in response to starting to use an ansi escape code in #6233.  Oddly, it doesn't check if its writing to a TTY...

Python's `tqdm` seems only check if its writing to a TTY ([code](https://github.com/tqdm/tqdm/blob/master/tqdm/std.py#L970))

As for Rust (generally in order of monthly downloads)
- [indicatif](https://lib.rs/crates/indicatif) only seems to check for TTY ([code](https://github.com/console-rs/indicatif/blob/16e755cd17570b95adff7d22c00c36b8014cd3e9/src/draw_target.rs#L128))
  - Side note: it doesn't seem to support ansi escape does on windows yet (console-rs/indicatif#388)
  - [tracing-indicatif](https://lib.rs/crates/tracing-indicatif) is a wrapper that automatically implements it based on `tracing`
- [prodash](https://lib.rs/crates/prodash) just checks if its writing to a TTY ([code](https://github.com/Byron/prodash/blob/main/src/render/line/draw.rs#L145))
  - `show_progress` is just used for delaying showing the progress bar, like we do with `cargo clean`
- [pbr](https://lib.rs/crates/pbr) doesn't seem to do detection
  - Potentially useful is some [windows API calls](https://github.com/a8m/pb/blob/master/src/tty/windows.rs) for when ANSI escape codes aren't supported
- [progress_bar](https://lib.rs/crates/progress_bar) has no detection
- [nutmeg](https://lib.rs/crates/nutmeg) only shows progress bars for TTYs where `!TERM=dumb` and supprts ansi escape codes (for windows) ([code](https://github.com/sourcefrog/nutmeg/blob/main/src/lib.rs#L923))
- [superconsole](https://lib.rs/crates/superconsole) only shows progress bars for TTYs where `!TERM=dumb` and supprts ansi escape codes (for windows) ([code](https://github.com/facebookincubator/superconsole/blob/62b189b0029da61d1b93dde6dac0af94f181a4aa/src/superconsole.rs#L82))
  - A reference is made to issues with Emacs' eshell but no other details are given
- [progressing](https://lib.rs/crates/progressing) doesn't do detection

Depending on how fancy we get, [unicode detection](https://crates.io/crates/supports-unicode) might become important

For reference on all of this, [cargo nextest](https://crates.io/crates/cargo-nextest/0.9.53/dependencies) is using `indicatif` and `supports-unicode`

I got a tip to check fish.  It seems to rely on terminfo with some hacks for Mac and Midnight Command ([code](https://github.com/fish-shell/fish-shell/blob/11c8d9684ee4d2e0a5b061a6c1b87e8842347cbc/fish-rust/src/env_dispatch.rs#L400))

On Discord, someone had this to say:
> i've done a lot of this. it's kind of a mess honestly. for anything advanced you need a bunch of heuristics, terminfo parsing, and DEC queries. that said for basic progress bar detection you should be able to just get by on terminfo parsing + heuristics to detect when terminfo might be lying
in general something like terminfo parsing + "treat any TERM with xterm in it as fairly capable" + "give up on any terminfo with delay expressions (e.g. $<blah>)" is probably reasonable

**Comment by bjorn3** [comment]
Another disadvantage of superconsole is that it breaks scrolling up to see the previously built crates (and in case of `-v` their rustc invocation) It also likely doesn't work well with `CARGO_LOG`.

**Comment by epage** [comment]
My expectation is that in using something like superconsole, we would stop listing all of the crates that built but that we would, like buck, provide a command to see what was previously built which would also help with #2904.

I think it would also be reasonable to revert back to the current progress system with `-v`.

**Comment by Byron** [comment]
I just wanted to add that `prodash` is merely used, by `gix`, as an interface to transmit hierarchical progress in potentially long-running operations. Even though it provides its own renderers I don't think they are suitable here so I presume even if `cargo` gets improved progress output, we would still stick with a translation layer like it's currently implemented, to display `git` specific progress.

`prodash` renderers definitely won't set the bar for understanding their terminal environment either, and when writing it "works for me" was good enough. It does respect [NO_COLOR](https://no-color.org) though.

**Comment by epage** [comment]
btw zig has made their output more dynamic and the HN thread has some discussion on the trade offs: https://news.ycombinator.com/item?id=40519976

**Comment by weihanglo** [comment]
> btw zig has made their output more dynamic and the HN thread has some discussion on the trade offs: [news.ycombinator.com/item?id=40519976](https://news.ycombinator.com/item?id=40519976)

A slightly different output style from the Nix community: <https://github.com/maralorn/nix-output-monitor>. (Also print a dependency graph.)

**Comment by epage** [comment]
As an update, we now have config settings for
- unicode
- hyperlinks

And soon, system progress reporting.

Those have been setting the pattern that the use of terminal capabilities is done on a terminal-by-terminal allow-list with an explicit control via config for when the allow list does not meet a users needs.  We can then extend that to richer terminal rendering than we have today.

**Comment by epage** [comment]
I realize I've never fully summarized my proposal for this on this issue.

My idea is that we designate output as "sticky" or "non-sticky":
- Sticky output, like warnings an errors, persists
- Non-sticky output (most status messages) will be cleared when their parent operation is done

I did say "most status messages", not "all".  My ideal "first step" would be for each command to have a "summary" status message that is sticky.  Due to how we compose things, this might initially be a "summary per step" (`cargo test` showing compile and test summaries) rather than a single "command summary"  (test summary).  Cargo scripts would have no run summary (*maybe* a compile summary *if* something was built).

We would also show multi-line progress indicators.  We'd want to survey how other multi-line progress indicators handle
- Very fast entries (flashing of output, order instability)
- More entries than can be shown (e.g. large builds on 64-core machines)
- How we can help people identify slow operations
- Can the progress bar report the phase of compilation (frontend, backend, link)?

Where supported, we should probably use the terminal sync control codes for nicer refreshes.

The way I picture this happening internally is that a status message can be parented under another status message.  One way of handling sticky items is to declare it independent of any parent status.  Maybe one way to handle "last status is sticky" is we just bake that assumption in?  Should we have implicit parenting of status messages by having a status stack in `GlobalContext`?

We should probably refactor the existing code to something more like a status system (e.g. instead of making direct calls to "clear line") and then work to allow support for a new status implementation underneath that.

**Comment by RalfJung** [comment]
For ui_test, @oli-obk built some pretty fancy multi-line status presentation, showing which tests are currently being run, and logging the ones that are completed above. I have no idea how it works but maybe the underlying tech could be useful for cargo, too.

**Comment by oli-obk** [comment]
I'm mostly just forwarding to indicatif, though with some buffering so that I don't add and remove items that get processed faster than a human can observe them.

One thing that is nicely possible in my setup is trees of builds and dependencies. So we could display what is essentially `cargo tree` where leaves get removed after they finish compilation. Though that can quickly get too large, so maybe just a list is actually better for cargo.

**Comment by Nemo157** [comment]
> So we could display what is essentially `cargo tree` where leaves get removed after they finish compilation. Though that can quickly get too large, so maybe just a list is actually better for cargo.

I prototyped something like this years ago; no idea if the code still works, but the asciinema should still be viewable: <https://github.com/nemo157/cargo-build-tree>. And you're right, it is way too large to be usable in most real projects—at least with  what I show there, but relooking at it now I think there are some simplifications of the tree that could be done which might make it closer to usable.

**Comment by hkalbasi** [comment]
Zig build progress bar might be a source of inspiration:

![Image](https://github.com/user-attachments/assets/11055eb7-82f7-4755-b20c-7714f59188c6)

A blog post explaining it: https://andrewkelley.me/post/zig-new-cli-progress-bar-explained.html

**Comment by Byron** [comment]
Seeing this I remember [this comment](https://github.com/rust-lang/cargo/issues/8889#issuecomment-2706645978) and believe it included something along the lines of "don't be overwhelming or flickery", and I think Cargo should try very hard to become the most ergonomic solution, one that just feels pleasant and "right" while giving just the information that users would be interested in.
This comment is not to start an argument, but to share some of the things I'd value, which also means I could value the Zig example more.

**Comment by Timmmm** [comment]
Agreed - I made #16388 for that specific problem, and I'll have a PR in a second (which works and seems reasonably clean to me but I'm not a cargo pro so I might have done something stupid).

**Comment by ehuss** [comment]
See also https://github.com/rust-lang/cargo/issues/7265 which asked for the console to show rustc's stage (like codegen). This is now available in rustc via the unstable [`--timings`](https://doc.rust-lang.org/nightly/rustc/json.html#timings) which cargo is using for some unstable things already.
